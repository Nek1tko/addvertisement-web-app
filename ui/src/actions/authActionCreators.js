import { SIGN_IN, SIGN_OUT } from "./types";

export const userSignIn = (phoneNumber, userId, token) => {
    return {
        type: SIGN_IN,
        payload: {
            phoneNumber: phoneNumber,
            userId: userId,
            token: token
        }
    };
};

export const userSignOut = () => {
    return {
        type: SIGN_OUT,
    };
};
