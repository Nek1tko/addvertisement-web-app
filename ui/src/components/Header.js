import React from "react";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import IconButton from "@material-ui/core/IconButton";
import MenuItem from "@material-ui/core/MenuItem";
import Menu from "@material-ui/core/Menu";
import Button from "@material-ui/core/Button";
import { withRouter } from "react-router-dom";
import { Avatar, Box, Tooltip } from "@material-ui/core";

const Header = props => {
    const { history } = props;

    const [anchorElUser, setAnchorElUser] = React.useState(null);

    const handleButtonClick = pageURL => {
        history.push(pageURL);
    };

    const handleOpenUserMenu = (event) => {
        setAnchorElUser(event.currentTarget);
    };

    const handleCloseUserMenu = pageURL => {
        setAnchorElUser(null);
        history.push(pageURL);
    };

    const homeItem = {
        menuTitle: "Главная",
        pageURL: "/"
    }
    const menuItems = [
        {
            menuTitle: "Мои объявления",
            pageURL: "/my-ads"
        },
        {
            menuTitle: "Создать объявление",
            pageURL: "/create-ad"
        }
    ];

    const settings = ['Профиль', 'Выйти'];
    const settingsItems = [
        {
            settingsTitle: "Профиль",
            pageURL: "/personal-area"
        },
        {
            settingsTitle: "Выйти",
            pageURL: "/login"
        }
    ];

    return (
        <AppBar style={{ background: '#669999' }} position="static">
            <Toolbar>
                <Button
                    style={{ color: '#FFFFFF' }}
                    onClick={() => handleButtonClick(homeItem.pageURL)} >
                    <Typography variant="h4" sx={{ flexGrow: 1 }}>
                        {homeItem.menuTitle}
                    </Typography>
                </Button>

                <Box sx={{ flexGrow: 1, display: 'flex', ml: 2 }}>
                    {menuItems.map(menuItem => {
                        const { menuTitle, pageURL } = menuItem;
                        return (
                            <Box sx={{ px: 2 }}>
                                <Button
                                    style={{ color: '#FFFFFF', textTransform: 'none', fontSize: '18px' }}
                                    onClick={() => handleButtonClick(pageURL)}
                                >
                                    {menuTitle}
                                </Button>
                            </Box>
                        );
                    })}
                </Box>

                <Box sx={{ flexGrow: 0 }}>
                    <Tooltip title="Открыть настройки">
                        <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                            <Avatar />
                        </IconButton>
                    </Tooltip>
                    <Menu
                        sx={{ mt: '45px' }}
                        id="menu-appbar3"
                        anchorEl={anchorElUser}
                        anchorOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        keepMounted
                        transformOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        open={Boolean(anchorElUser)}
                        onClose={handleCloseUserMenu}
                    >
                        {settingsItems.map((settingsItem) => {
                            const { settingsTitle, pageURL } = settingsItem;
                            return (
                                <MenuItem key={settingsTitle} onClick={() => handleCloseUserMenu(pageURL)}>
                                    <Typography>{settingsTitle}</Typography>
                                </MenuItem>
                            );
                        })}
                    </Menu>
                </Box>
            </Toolbar>
        </AppBar >
    );
};

export default withRouter(Header);
