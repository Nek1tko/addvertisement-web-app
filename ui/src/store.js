import { createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import authReducer from "./reducers/authReducer";

const middleware = [thunk];

const store = createStore(
    authReducer,
    composeWithDevTools(applyMiddleware(...middleware))
);

export default store;