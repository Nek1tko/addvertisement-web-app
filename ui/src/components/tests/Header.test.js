import { mount, shallow } from 'enzyme';
import Header from '../Header';
import { BrowserRouter, Router } from 'react-router-dom';
import AuthService from '../../services/auth.service';
import { createBrowserHistory } from 'history'
import { userSignIn, userSignOut } from '../../actions/authActionCreators';
import store from '../../store';
import { act } from 'react-dom/test-utils';

describe('HeaderTests', () => {
    let mHistory, wrapper

    beforeEach(() => {
        mHistory = {
            ...createBrowserHistory(),
            push: jest.fn()
        };
        wrapper = mount(
            <Router history={mHistory}>
                <Header />
            </Router>
        );
    })

    afterEach(() => {
        act(() => {
            store.dispatch(userSignOut());
        })
    })

    test('HeaderRenderTest', () => {
        mount(
            <BrowserRouter>
                <Header />
            </BrowserRouter>
        );
    });

    test('HeaderSnapshotTest', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('HeaderHomeMenuButtonsTest', () => {
        const homeButton = wrapper.find('[id="homeButton"]').first()
        homeButton.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/");
    });

    test('HeaderMyAdsMenuButtonsTest', () => {
        const myAdsButton = wrapper.find('[id="Мои объявления"]').first()
        myAdsButton.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/my-ads");
    });

    test('HeaderCreateAdMenuButtonsTest', () => {
        const createAdButton = wrapper.find('[id="Создать объявление"]').first()
        createAdButton.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/create-ad");
    });

    test('HeaderSettingsCollapseTest', () => {
        const mUser = {
            userId: 0,
            phoneNumber: "phone",
            token: "token"
        };

        const settingsCollapse = wrapper.find('[id="settingsCollapse"]').first();
        expect(settingsCollapse.prop("in")).toBe(false);

        act(() => {
            store.dispatch(userSignIn(mUser.phoneNumber, mUser.userId, mUser.token));
            wrapper.update();
        });

        const unsubscribe = store.subscribe(() => {
            expect(settingsCollapse.prop("in")).toBe(true);
        })
        unsubscribe()
    });

    test('HeaderSettingsIconButtonTest', () => {
        const settingsMenu = wrapper.find('[id="menu-appbar3"]').first();
        expect(settingsMenu.prop("anchorEl")).toBe(null)

        const settingsIconButton = wrapper.find('[id="settingsIconButton"]').first();
        settingsIconButton.simulate("click");

        // smthg odd
        expect(settingsMenu.prop("anchorEl")).toBe(null)
    });

    test('HeaderPersonalAreaSettingsMenuItemTest', () => {
        const personalAreaSettingsMenuItem = wrapper.find('[id="Профиль"]').first();
        personalAreaSettingsMenuItem.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/personal-area");
    });

    test('HeaderFavoritesSettingsMenuItemTest', () => {
        const favoritesSettingsMenuItem = wrapper.find('[id="Избранное"]').first();
        favoritesSettingsMenuItem.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/favorites");
    });

    test('HeaderSignOutSettingsMenuItemTest', () => {
        const mAuthServiceSignOut = jest.spyOn(AuthService, "signOut");
        mAuthServiceSignOut.mockResolvedValue();

        const signOutSettingsItem = wrapper.find('[id="Выйти"]').first();
        signOutSettingsItem.simulate('click');

        expect(mHistory.push).toHaveBeenCalledTimes(1);
        expect(mHistory.push).toBeCalledWith("/login");
        expect(mAuthServiceSignOut).toHaveBeenCalledTimes(1);
    });
});
