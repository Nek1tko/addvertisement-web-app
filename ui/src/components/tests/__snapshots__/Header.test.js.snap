// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderTests HeaderSnapshotTest 1`] = `ReactWrapper {}`;
